File: archive/llm_written_v1/gpt-5/smith-waterman/c#/algo_parallel.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace SmithWatermanAlignment
{
    // Parallel implementation using tiled anti-diagonal (wavefront) parallelism
    public class SmithWaterman
    {
        private readonly int matchScore;
        private readonly int mismatchScore;
        private readonly int gapScore;
        // Sequential fallback threshold (cells). Conservative to avoid overhead on small problems.
        private readonly int smallThresholdCells = 1_000_000; // 1M cells
        private readonly int tile = 128; // tile size for blocked wavefront

        public SmithWaterman(int match, int mismatch, int gap)
        {
            matchScore = match;
            mismatchScore = mismatch;
            gapScore = gap;
        }

        public int[][] ConstructMatrix(string query, string reference)
        {
            int n = query.Length + 1; // rows
            int m = reference.Length + 1; // cols

            // Use dense 2D array for better locality; convert to jagged at end
            int[,] H2 = new int[n, m];

            // Cache string data to local char arrays for faster repeated access
            var qArr = query.ToCharArray();
            var rArr = reference.ToCharArray();

            int cells = (n - 1) * (m - 1);
            int maxWorkers = Math.Max(1, Environment.ProcessorCount);
            var opts = new ParallelOptions { MaxDegreeOfParallelism = maxWorkers };

            // Sequential fast path for small inputs or single core
            if (cells <= smallThresholdCells || maxWorkers == 1)
            {
                for (int i = 1; i < n; i++)
                {
                    for (int j = 1; j < m; j++)
                    {
                        int scoreDiagonal = H2[i - 1, j - 1] + (qArr[i - 1] == rArr[j - 1] ? matchScore : mismatchScore);
                        int scoreUp = H2[i - 1, j] + gapScore;
                        int scoreLeft = H2[i, j - 1] + gapScore;
                        int best = scoreDiagonal;
                        if (scoreUp > best) best = scoreUp;
                        if (scoreLeft > best) best = scoreLeft;
                        if (best < 0) best = 0;
                        H2[i, j] = best;
                    }
                }
            }
            else
            {
                // Blocked wavefront over tiles
                int B = tile;
                int nBlocks = (n - 1 + B - 1) / B; // number of blocks covering rows 1..n-1
                int mBlocks = (m - 1 + B - 1) / B; // covering cols 1..m-1

                // Iterate tile anti-diagonals
                for (int td = 0; td <= (nBlocks - 1) + (mBlocks - 1); td++)
                {
                    int biStart = Math.Max(0, td - (mBlocks - 1));
                    int biEnd = Math.Min(nBlocks - 1, td);
                    int len = biEnd - biStart + 1;
                    if (len <= 0) continue;

                    Parallel.For(0, len, opts, t =>
                    {
                        int bi = biStart + t;
                        int bj = td - bi;
                        int i0 = 1 + bi * B;
                        int i1 = Math.Min(n - 1, i0 + B - 1);
                        int j0 = 1 + bj * B;
                        int j1 = Math.Min(m - 1, j0 + B - 1);

                        // Compute this tile sequentially (respecting in-tile dependencies)
                        for (int i = i0; i <= i1; i++)
                        {
                            int qi = i - 1;
                            for (int j = j0; j <= j1; j++)
                            {
                                int scoreDiagonal = H2[i - 1, j - 1] + (qArr[qi] == rArr[j - 1] ? matchScore : mismatchScore);
                                int scoreUp = H2[i - 1, j] + gapScore;
                                int scoreLeft = H2[i, j - 1] + gapScore;
                                int best = scoreDiagonal;
                                if (scoreUp > best) best = scoreUp;
                                if (scoreLeft > best) best = scoreLeft;
                                if (best < 0) best = 0;
                                H2[i, j] = best;
                            }
                        }
                    });
                    // Barrier at end of Parallel.For per tile diagonal
                }
            }

            // Convert to jagged array for external API compatibility
            int[][] H = new int[n][];
            for (int i = 0; i < n; i++)
            {
                H[i] = new int[m];
                for (int j = 0; j < m; j++) H[i][j] = H2[i, j];
            }

            return H;
        }

        public (int, int) FindHighestScore(int[][] H)
        {
            int maxScore = 0;
            int maxI = 0;
            int maxJ = 0;

            for (int i = 0; i < H.Length; i++)
            {
                for (int j = 0; j < H[i].Length; j++)
                {
                    if (H[i][j] > maxScore)
                    {
                        maxScore = H[i][j];
                        maxI = i;
                        maxJ = j;
                    }
                }
            }

            return (maxI, maxJ);
        }

        public (string, string, int, double) Traceback(int[][] H, string query, string reference)
        {
            List<char> alignedA = new List<char>();
            List<char> alignedB = new List<char>();

            var (i, j) = FindHighestScore(H);
            int score = H[i][j];

            int totalMatch = 0;
            int totalAlignment = 0;

            while (i > 0 && j > 0)
            {
                int currentScore = H[i][j];
                if (currentScore == 0) break;

                int diagonalScore = H[i - 1][j - 1];
                int upScore = H[i - 1][j];
                int leftScore = H[i][j - 1];

                int expectedDiagonal = diagonalScore + (query[i - 1] == reference[j - 1] ? matchScore : mismatchScore);

                if (currentScore == expectedDiagonal)
                {
                    alignedA.Add(query[i - 1]);
                    alignedB.Add(reference[j - 1]);
                    totalAlignment++;
                    if (query[i - 1] == reference[j - 1]) totalMatch++;
                    i--; j--;
                }
                else if (currentScore == upScore + gapScore)
                {
                    alignedA.Add(query[i - 1]);
                    alignedB.Add('-');
                    totalAlignment++;
                    i--;
                }
                else if (currentScore == leftScore + gapScore)
                {
                    alignedA.Add('-');
                    alignedB.Add(reference[j - 1]);
                    totalAlignment++;
                    j--;
                }
                else
                {
                    break;
                }
            }

            alignedA.Reverse();
            alignedB.Reverse();

            double percentageIdentity = (totalAlignment > 0) ? ((double)totalMatch / totalAlignment) * 100.0 : 0.0;

            string alignedAString = new string(alignedA.ToArray());
            string alignedBString = new string(alignedB.ToArray());

            return (alignedAString, alignedBString, score, percentageIdentity);
        }

        public (string, string, int, double) FindAlignment(string query, string reference)
        {
            int[][] H = ConstructMatrix(query, reference);
            return Traceback(H, query, reference);
        }
    }
}

File: archive/llm_written_v1/gpt-5/smith-waterman/cpp/algo_parallel.cpp
#include "smith_waterman.h"
#include <algorithm>
#include <cstddef>
#include <vector>
#include <string>
#include <tuple>
#include <utility>
#include <cstdlib>

#ifdef _OPENMP
#include <omp.h>
#endif

// Helper: clamp threads to available cores (if OpenMP present)
static int bounded_thread_count() {
#ifdef _OPENMP
    int procs = omp_get_num_procs();
    int max_t = omp_get_max_threads();
    if (procs <= 0) procs = 1;
    if (max_t <= 0) max_t = 1;
    int t = procs < max_t ? procs : max_t;
    if (t < 1) t = 1;
    return t;
#else
    return 1;
#endif
}

SmithWaterman::SmithWaterman(int match, int mismatch, int gap)
    : matchScore(match), mismatchScore(mismatch), gapScore(gap) {}

std::vector<std::vector<int>> SmithWaterman::constructMatrix(
    const std::string& query,
    const std::string& reference) {

    const int n = static_cast<int>(query.length()) + 1;
    const int m = static_cast<int>(reference.length()) + 1;

    std::vector<std::vector<int>> H(n, std::vector<int>(m, 0));

    // Early exits and tiny-input sequential fast path
    if (n <= 1 || m <= 1) {
        return H;
    }

    // Total DP cells excluding the first row/col
    const long long work = static_cast<long long>(n - 1) * static_cast<long long>(m - 1);

    // Threshold for switching to parallel wavefront (avoid overhead on tiny cases)
    const long long PAR_THRESHOLD = 20000; // ~20k cells

    int threads = bounded_thread_count();

    bool do_parallel = (work >= PAR_THRESHOLD) && (threads > 1);

#ifndef _OPENMP
    do_parallel = false;
#endif

    if (!do_parallel) {
        // Sequential baseline fill (row-major)
        for (int i = 1; i < n; ++i) {
            for (int j = 1; j < m; ++j) {
                int scoreDiagonal = H[i-1][j-1] + (query[i-1] == reference[j-1] ? matchScore : mismatchScore);
                int scoreUp = H[i-1][j] + gapScore;
                int scoreLeft = H[i][j-1] + gapScore;
                H[i][j] = std::max({0, scoreDiagonal, scoreUp, scoreLeft});
            }
        }
        return H;
    }

#ifdef _OPENMP
    // Parallel wavefront fill along anti-diagonals (k = i + j)
    // Dependencies: each cell depends on (i-1,j-1), (i-1,j), (i,j-1) ∈ previous anti-diagonals.
    // We rely on the implicit barrier at the end of each omp for to separate diagonals.
    omp_set_num_threads(threads);

#pragma omp parallel default(none) shared(H, query, reference, n, m)
    {
        for (int k = 2; k <= (n - 1) + (m - 1); ++k) {
            int i_start = std::max(1, k - (m - 1));
            int i_end = std::min(n - 1, k - 1);
#pragma omp for schedule(static)
            for (int i = i_start; i <= i_end; ++i) {
                int j = k - i;
                // Safe because all dependencies are on diagonal k-1
                int scoreDiagonal = H[i-1][j-1] + (query[i-1] == reference[j-1] ? matchScore : mismatchScore);
                int scoreUp = H[i-1][j] + gapScore;
                int scoreLeft = H[i][j-1] + gapScore;
                H[i][j] = std::max({0, scoreDiagonal, scoreUp, scoreLeft});
            }
            // implicit barrier here (end of omp for) ensures diagonal order
        }
    }
#endif

    return H;
}

std::pair<int, int> SmithWaterman::findHighestScore(
    const std::vector<std::vector<int>>& H) {

    int maxScore = 0;
    std::pair<int, int> maxPos = {0, 0};

    for (int i = 0; i < static_cast<int>(H.size()); ++i) {
        for (int j = 0; j < static_cast<int>(H[i].size()); ++j) {
            int v = H[i][j];
            if (v > maxScore) {
                maxScore = v;
                maxPos = {i, j};
            }
        }
    }

    return maxPos;
}

std::tuple<std::string, std::string, int, double> SmithWaterman::traceback(
    const std::vector<std::vector<int>>& H,
    const std::string& query,
    const std::string& reference) {

    std::string alignedA;
    std::string alignedB;

    auto [i, j] = findHighestScore(H);
    int score = H[i][j];

    int totalMatch = 0;
    int totalAlignment = 0;

    while (i > 0 && j > 0) {
        int currentScore = H[i][j];
        if (currentScore == 0) break;

        int diagonalScore = H[i-1][j-1];
        int upScore = H[i-1][j];
        int leftScore = H[i][j-1];

        int expectedDiagonal = diagonalScore + (query[i-1] == reference[j-1] ? matchScore : mismatchScore);

        if (currentScore == expectedDiagonal) {
            alignedA.push_back(query[i-1]);
            alignedB.push_back(reference[j-1]);
            ++totalAlignment;
            if (query[i-1] == reference[j-1]) ++totalMatch;
            --i; --j;
        } else if (currentScore == upScore + gapScore) {
            alignedA.push_back(query[i-1]);
            alignedB.push_back('-');
            ++totalAlignment;
            --i;
        } else if (currentScore == leftScore + gapScore) {
            alignedA.push_back('-');
            alignedB.push_back(reference[j-1]);
            ++totalAlignment;
            --j;
        } else {
            break;
        }
    }

    std::reverse(alignedA.begin(), alignedA.end());
    std::reverse(alignedB.begin(), alignedB.end());

    double percentageIdentity = (totalAlignment > 0)
        ? (static_cast<double>(totalMatch) / static_cast<double>(totalAlignment)) * 100.0
        : 0.0;

    return {alignedA, alignedB, score, percentageIdentity};
}

std::tuple<std::string, std::string, int, double> SmithWaterman::findAlignment(
    const std::string& query,
    const std::string& reference) {

    auto H = constructMatrix(query, reference);
    return traceback(H, query, reference);
}
